crates/executor/src/blockchain_tree/mod.rs

```diff
@@ -0,0 +1,811 @@
+        };
+        let chain = self.chains.remove(&chain_id).expect("To be present");
+
+        // we are spliting chain as there is possibility that only part of chain get canonicalized.
+        let (canonical, pending) = chain.split_at_block_hash(block_hash);
+        let canonical = canonical.expect("Canonical chain is present");
+
+        if let Some(pending) = pending {
+            // fork is now canonical and latest.
+            self.block_indices.insert_chain(chain_id, &pending);
+            self.chains.insert(chain_id, pending);
+        }
+
+        let mut block_fork = canonical.fork_block();
+        let mut block_fork_number = canonical.fork_block_number();
+        let mut chains_to_promote = vec![canonical];
+
+        // loop while fork blocks are found in Tree.
+        while let Some(chain_id) = self.block_indices.get_blocks_chain_id(&block_fork.hash) {
+            let chain = self.chains.remove(&chain_id).expect("To fork to be present");
+            block_fork = chain.fork_block();
+            let (canonical, rest) = chain.split_at_number(block_fork_number);
+            let canonical = canonical.expect("Chain is present");
+            // reinsert back the chunk of sidechain that didn't get reorged.
+            if let Some(rest_of_sidechain) = rest {
+                self.block_indices.insert_chain(chain_id, &rest_of_sidechain);
+                self.chains.insert(chain_id, rest_of_sidechain);
+            }
+            block_fork_number = canonical.fork_block_number();
+            chains_to_promote.push(canonical);
+        }
+
+        let old_tip = self.block_indices.canonical_tip();
+        // Merge all chain into one chain.
+        let mut new_canon_chain = chains_to_promote.pop().expect("There is at least one block");
+        for chain in chains_to_promote.into_iter().rev() {
+            new_canon_chain.append_chain(chain);
+        }
+
+        // update canonical index
+        self.block_indices.canonicalize_blocks(&new_canon_chain.blocks);
+
+        // if joins to the tip
+        if new_canon_chain.fork_block_hash() == old_tip.hash {
+            // append to database
+            self.commit_canonical(new_canon_chain)?;
+        } else {
+            // it forks to canonical block that is not the tip.
+
+            let canon_fork = new_canon_chain.fork_block();
+            // sanity check
+            if self.block_indices.canonical_hash(&canon_fork.number) != Some(canon_fork.hash) {
+                unreachable!("all chains should point to canonical chain.");
+            }
+
+            // revert `N` blocks from current canonical chain and put them inside BlockchanTree
+            // This is main reorgs on tables.
+            let old_canon_chain = self.revert_canonical(canon_fork.number)?;
+            self.commit_canonical(new_canon_chain)?;
+
+            // insert old canonical chain to BlockchainTree.
+            // TODO check if there there is chains that can be merged.
```

Comment:
Is this still a todo?

Reply:
Left it there as potential optimization hint, reverted canonical chain can potentially be extended with one of the sidechains, we can go without it.

Reworded todo